File: kubi-api/src/routes/stream.ts
import { Hono } from "hono";
import { z } from "zod";
import { callKubiStream } from "../lib/client";

const streamSchema = z.object({
  query: z.string().min(1).max(2000),
  type: z.enum(["instant", "fast", "auto", "deep-lite", "deep", "research"]).default("auto"),
  numResults: z.number().int().min(1).max(100).default(10),
  includeText: z.boolean().default(false),
  systemPrompt: z.string().optional(),
});

export const streamRoutes = new Hono();

streamRoutes.post("/", async (c) => {
  const raw = await c.req.json();
  const parsed = streamSchema.safeParse(raw);
  if (!parsed.success) return c.json({ error: "Validation error", details: parsed.error.flatten() }, 400);

  const d = parsed.data;
  try {
    const aiResponse = await callKubiStream({ path: "/ai/stream", body: {
      query: d.query, type: d.type, num_results: d.numResults,
      include_text: d.includeText, system_prompt: d.systemPrompt,
    }});

    return new Response(aiResponse.body, {
      headers: { "Content-Type": "text/event-stream", "Cache-Control": "no-cache", Connection: "keep-alive", "X-Accel-Buffering": "no" },
    });
  } catch (err: any) {
    return c.json({ error: "Stream failed", message: err.message }, 500);
  }
});

File: kubi-core/src/routes/stream.rs
use axum::{body::Body, extract::State, response::{IntoResponse, Response}, Json};
use serde_json::json;
use crate::AppState;

pub async fn stream(State(state): State<AppState>, Json(body): Json<serde_json::Value>) -> impl IntoResponse {
    let url = format!("{}/ai/stream", state.kubi_ai_url);
    match state.http_client.post(&url).json(&body).send().await {
        Ok(resp) => {
            let status = resp.status();
            let body = Body::from_stream(resp.bytes_stream());
            match Response::builder()
                .status(status)
                .header("content-type", "text/event-stream")
                .header("cache-control", "no-cache")
                .header("connection", "keep-alive")
                .body(body)
            {
                Ok(r) => r,
                Err(e) => {
                    let err_body = Body::from(json!({"error": "Stream build failed", "message": e.to_string()}).to_string());
                    Response::builder().status(500).header("content-type", "application/json").body(err_body).unwrap()
                }
            }
        }
        Err(e) => {
            tracing::error!("Stream failed: {}", e);
            let body = Body::from(json!({"error": "Stream failed", "message": e.to_string()}).to_string());
            Response::builder().status(500).header("content-type", "application/json").body(body).unwrap()
        }
    }
}
